signalingserver.go

```diff
@@ -14,23 +14,161 @@
 // forwarding all requests to it.
-type ForwardingSignalingServer struct{}
+type ForwardingSignalingServer struct {
+	instanceManager InstanceManager
+}
 
-func NewForwardingSignalingServer() *ForwardingSignalingServer {
-	return &ForwardingSignalingServer{}
+func NewForwardingSignalingServer(im InstanceManager) *ForwardingSignalingServer {
+	return &ForwardingSignalingServer{im}
 }
 
-func (s *ForwardingSignalingServer) NewConnection(msg NewConnMsg) NewConnReply {
-	return NewConnReply{ConnId: "id", DeviceInfo: "device info"}
+func (s *ForwardingSignalingServer) NewConnection(msg NewConnMsg) (SServerResponse, error) {
+	dev, err := s.instanceManager.DeviceFromId(msg.DeviceId)
+	if err != nil {
+		return SServerResponse{StatusCode: http.StatusInternalServerError}, err
+	}
+	var resErr ErrorMsg
+	var reply NewConnReply
+	status, err := POSTRequest(hostURL(dev.Addr, "/polled_connections", ""), NewConnMsg{dev.LocalId}, &reply, &resErr)
+	if err != nil {
+		return SServerResponse{StatusCode: status}, err
+	}
+	if resErr.Error != "" {
+		log.Println("The device host returned an error: ", resErr.Error)
+		return SServerResponse{Response: resErr, StatusCode: status}, nil
+	}
+	// Add the device id to the connection id to reference it in future calls
+	reply.ConnId = encodeConnId(reply.ConnId, msg.DeviceId)
+	return SServerResponse{Response: reply, StatusCode: status}, nil
+}
+
+func (s *ForwardingSignalingServer) Forward(id string, msg ForwardMsg) (SServerResponse, error) {
+	dec, err := decodeConnId(id)
+	if err != nil {
+		log.Println("Failed to decode connection id: ", err)
+		return SServerResponse{StatusCode: http.StatusBadRequest}, err
+	}
+	connId := dec.ConnId
+	dev, err := s.instanceManager.DeviceFromId(dec.DevId)
+	if err != nil {
+		return SServerResponse{StatusCode: http.StatusInternalServerError}, err
+	}
+	var resErr ErrorMsg
+	var reply interface{}
+	status, err := POSTRequest(hostURL(dev.Addr, "/polled_connections/"+connId+"/:forward", ""), msg, &reply, &resErr)
+	return SServerResponse{reply, status}, err
+}
+
+func (s *ForwardingSignalingServer) Messages(id string, start int, count int) (SServerResponse, error) {
+	dec, err := decodeConnId(id)
+	if err != nil {
+		log.Println("Failed to decode connection id: ", err)
+		return SServerResponse{StatusCode: http.StatusBadRequest}, err
+	}
+	connId := dec.ConnId
+	dev, err := s.instanceManager.DeviceFromId(dec.DevId)
+	if err != nil {
+		return SServerResponse{StatusCode: http.StatusInternalServerError}, err
+	}
+	query := fmt.Sprintf("start=%d", start)
+	if count > 0 {
+		query = fmt.Sprintf("%s&count=%d", query, count)
+	}
+	var resErr ErrorMsg
+	var reply interface{}
+	status, err := GETRequest(hostURL(dev.Addr, "/polled_connections/"+connId+"/messages", query), &reply, &resErr)
+	return SServerResponse{reply, status}, err
+}
+
+func hostURL(addr string, path string, query string) string {
+	url := "http://" + addr + ":1080" + path
```

Comment:
The android-cuttlefish project promises this port number is [configurable](https://github.com/google/android-cuttlefish/blob/0861e8d9e6f2dee7f6070fb20dde98d7eb47dce4/debian/cuttlefish-common.default#L42) but here it's hardcoded, is this temporary?

Reply:
Yes, this is temporary. I need to figure out what's the best way to configure the app. That configuration will include the port used by the host orchestrator, the protocol and possible some auth metadata.